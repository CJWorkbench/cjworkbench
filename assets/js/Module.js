/**
 * A component that holds a single module that is then contained within the
 * Module Library.
 * The render function here will drive the "card" of each module within
 * the library.
 */

import React from 'react'
import PropTypes from 'prop-types'
import { CardBlock, Card } from 'reactstrap';

export default class Module extends React.Component {
  constructor(props) {
    super(props);
    this.initFields(props);
    this.state = {
      name: props.key,
      description: props.description,
      category: props.category,
      author: props.author,
    };
    this.itemClick = this.itemClick.bind(this);
    this.addModule = this.props.addModule.bind(this);
    this.workflow = this.props.workflow;
  }

  initFields(props) {
    this.key = props['data-name'];
    this.name = props['data-name'];
    this.description = props['data-description'];
    this.category = props['data-category'];
    this.author = props['data-author'];
    this.id = props['data-id'];
  }

  itemClick(evt) {
    var itemID = evt.target.getAttribute('data-id');
    this.props.addModule(this.props['data-id']);
    // Toggle temporarily disabled
    // this.workflow.toggleModuleLibrary();
  }

  render() {
    var moduleName = this.props['data-name'];
    var icon = 'icon-' + this.props['icon'] + ' ml-icon';

    return (
      // TODO: remove inline styles
      <div className='card' style={{'borderRadius': 0, 'border': 0}}>
        <div className='' onClick={this.itemClick} >
          <div className='second-level d-flex flex-row align-items-center'>
            <div className='ml-icon-container'>
              <div className={icon}></div>
            </div>
            <div>
              <div className='content-5 ml-module-name'>{moduleName}</div>
            </div>
          </div>
        </div>
      </div>
    )
  }
}

Module.propTypes = {
  addModule:  PropTypes.func,
  workflow: PropTypes.object,
};
