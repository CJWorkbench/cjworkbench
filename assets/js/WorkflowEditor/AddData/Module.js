import React from 'react'
import PropTypes from 'prop-types'

export default function Module ({ isLessonHighlight, idName, name, description, icon, onClick }) {
  const handleClick = React.useCallback(ev => { ev.preventDefault(); onClick(idName) })

  let className = 'module'
  if (isLessonHighlight) className += ' lesson-highlight'

  return (
    <a href='#' name={idName} onClick={handleClick} className={className}>
      <i className={`icon-${icon}`}/>
      <div className='text'>
        <h4>{name}</h4>
        <p>{description}</p>
      </div>
    </a>
  )
}
Module.propTypes = {
  idName: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  icon: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired // func(idName) => undefined
}
