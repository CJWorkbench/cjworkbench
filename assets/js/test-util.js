import { Provider } from 'react-redux'
import React from 'react'
import { shallow } from 'enzyme'

/**
 * Like enzyme's `shallow()`, but child components that depend on redux
 * don't crash the mount.
 *
 * See https://github.com/airbnb/enzyme/issues/472
 */
export function shallowWithStubbedStore(component) {
  const stub = () => ({})
  const store = { getState: stub, subscribe: stub, dispatch: stub }
  return shallow(<Provider store={store}>{component}</Provider>).dive()
}
