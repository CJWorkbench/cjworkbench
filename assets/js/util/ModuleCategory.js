import React from 'react'
import { t } from '@lingui/macro'
import { withI18n } from '@lingui/react'

const CategoryNames = {
  Combine: t('js.util.ModuleCategory.CategoryNames.Combine')`Combine`,
  Scrape: t('js.util.ModuleCategory.CategoryNames.Scrape')`Scrape`,
  Clean: t('js.util.ModuleCategory.CategoryNames.Clean')`Clean`,
  Analyze: t('js.util.ModuleCategory.CategoryNames.Analyze')`Analyze`,
  Visualize: t('js.util.ModuleCategory.CategoryNames.Visualize')`Visualize`,
  Code: t('js.util.ModuleCategory.CategoryNames.Code')`Code`,
  'Add data': t('js.util.ModuleCategory.CategoryNames.AddData')`Add data`,
  Other: t('js.util.ModuleCategory.CategoryNames.Other')`Other`
}

export const CategoryName = withI18n()(({ category, i18n }) => {
  return <>{getCategoryName(category, i18n)}</>
})

/**
 * When you need the category name as a string, you may use this function
 */
export function getCategoryName (category, i18n) {
  if (CategoryNames[category]) return i18n._(CategoryNames[category])
  else throw new Error('No such category: ' + category)
}
