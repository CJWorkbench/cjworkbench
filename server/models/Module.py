# A Module defines a data processing action, with a set of paramters
# Module is the class, WfModule is the instance (applied in a Workflow)
from django.db import models

class Module(models.Model):
    # UI name, can change
    name = models.CharField('name', max_length=200)

    # internal name, cannot change if you want backwards compatibility with exported workflows
    id_name = models.CharField('id_name', max_length=200)

    # how do we run this module?
    dispatch = models.CharField('dispatch', max_length=200)

    def __str__(self):
        return self.name