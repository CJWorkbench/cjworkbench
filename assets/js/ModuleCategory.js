/**
 * A component that holds a collection of modules for a given category. For example,
 * Category: Source
 * Modules: Load URL, Paste CSV, Twitter
 *
 * Category: Wrangle
 * Modules: Melt, Select Columns
 *
 * Categories should be expandable and collapsible, just like each individual module.
 */

import React from 'react'
import PropTypes from 'prop-types'
import Module from './Module'
import { Collapse, Button, CardBlock, Card } from 'reactstrap';

export default class ModuleCategory extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      collapsed: props.collapsed,
    };

    this.toggleCollapse = this.toggleCollapse.bind(this)
  }

  toggleCollapse() {
    var newCollapsed = !this.state.collapsed;
    this.setState({collapsed: newCollapsed});
    if (!newCollapsed) {
      this.props.setOpenCategory(this.props.name); // tell parent, so it can close other cats
    }
  }

  // When our props change, update our collapsed state (this is the other end of setOpenCategory)
  componentWillReceiveProps(newProps) {
    this.setState({collapsed: newProps.collapsed})
  }

  render() {
    var isOpen = !this.state.collapsed;

    // Provides margins around opened library category
    var cardClass = isOpen
      ? 'card b-l-gray library-card-category-open'
      : 'card b-l-gray library-card-category-closed';

    var symbol = isOpen
      ? 'icon-sort-down-vl-gray ml-sort button-icon-library ml-3 mb-1'
      : 'icon-sort-right-vl-gray button-icon-library ml-3';

    // --- Need a mapping of category-to-icon before implementing
    // var icon = 'icon-' + ??? + ' button-icon mr-2';

    return (
      <div className={cardClass}>
        <div className="cat-open">
          <div className='first-level d-flex align-items-center'onClick={this.toggleCollapse}>
            <div className='cat-container'>
              <div className={symbol} />
              {/* <span className={icon}></span> */}
              <span className='content-3 t-vl-gray ml-3 ml-cat'>{this.props.name}</span>
            </div>
          </div>
          <div>
            <Collapse className='' isOpen={isOpen}>
              <div className="ml-list">{this.props.modules}</div>
            </Collapse>
          </div>
        </div>
      </div>
    );
  }
}


ModuleCategory.propTypes = {
  name:             PropTypes.string.isRequired,
  modules:          PropTypes.arrayOf(PropTypes.object).isRequired,
  collapsed:        PropTypes.bool.isRequired,
  setOpenCategory:  PropTypes.func.isRequired,
  isReadOnly:       PropTypes.bool.isRequired
};
