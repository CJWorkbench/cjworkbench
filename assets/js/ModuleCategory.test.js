/**
 * Testing Stories:
 * -Renders library-open version
 *    -Clicking on category will toggle collapse of module list
 * -Renders library-closed version
 *    -Mouse enter on category will show module list
 *    -Mouse leave on category will hide module list
 *    -Mouse leave on module list will hide module list
 *
 */

import React from 'react'
import ModuleCategory from './ModuleCategory'
import Module from './Module'
import { shallowWithStubbedStore } from './test-util'

describe('ModuleCategory ', () => {
  let wrapper
  let setOpenCategory
  const modules = [
    { id: 88, icon: 'add', name: 'First Module' },
    { id: 101, icon: 'url', name: 'Second Module' },
  ]

  beforeEach(function() {
    setOpenCategory = jest.fn()
  })

  describe('Library open', () => {
    beforeEach(() => {
      wrapper = shallowWithStubbedStore(
        <ModuleCategory
          name={"Add Data"}
          modules={modules}
          isReadOnly={false}
          setOpenCategory={setOpenCategory}
          libraryOpen={true}
          collapsed={false}
          />
      )
    })

    it('matches snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('Renders with list of Module components', () => {
      expect(wrapper.find(Module)).toHaveLength(2)
    })
  })

  describe('Library closed, category collapsed ', () => {
    beforeEach(() => {
      wrapper = shallowWithStubbedStore(
        <ModuleCategory
          name={"Add data"}
          modules={modules}
          isReadOnly={false}
          collapsed={true}
          setOpenCategory={setOpenCategory}
          libraryOpen={false}
          />
      )
    })

    it('matches snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    })

    it('renders with category icon, but without list of Module components', () => {
      // check that correct icon is displayed for "Add Data"
      expect(wrapper.find('.icon-database')).toHaveLength(1)
      // check for absence of Modules
      expect(wrapper.find('.ml-module-card')).toHaveLength(0)
    })

    it('opens module list on category mouseenter', () => {
      // find category card
      let category = wrapper.find('.ML-cat')
      expect(category).toHaveLength(1)
      // ensure absence of module list
      let moduleList = wrapper.find('.ml-list-mini')
      expect(moduleList).toHaveLength(0)
      // mouse enters category
      category.simulate('mouseEnter')
      // check: setOpenCategory() called from props
      expect(setOpenCategory.mock.calls.length).toBe(1)
    })
  })

  describe('Library closed, category open', () => {
    beforeEach(() => {
      wrapper = shallowWithStubbedStore(
        <ModuleCategory
          name={"Add Data"}
          modules={modules}
          isReadOnly={false}
          collapsed={false}
          setOpenCategory={setOpenCategory}
          libraryOpen={false}
        />
      )
    })

    it('matches snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    })

    it('renders with list of Module components', () => {
      expect(wrapper.find(Module)).toHaveLength(2)
    })

    it('closes module list on category mouseleave', () => {
      // find category card
      const category = wrapper.find('.ML-cat')
      expect(category).toHaveLength(1)
      expect(wrapper.find('.ml-list-mini')).toHaveLength(1)
      category.simulate('mouseLeave')
      expect(setOpenCategory).toHaveBeenCalledWith(null)
    })

    it('closes module list on .ml-list-mini mouseleave', () => {
      let moduleList = wrapper.find('.ml-list-mini')
      moduleList.simulate('mouseLeave')
      expect(setOpenCategory).toHaveBeenCalledWith(null)
    })
  })
})
