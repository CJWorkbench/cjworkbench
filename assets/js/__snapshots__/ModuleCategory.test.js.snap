// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleCategory  Library closed, category collapsed  matches snapshot 1`] = `
<div
  className="card closed-ML--category"
>
  <div
    className="module-category--wrapper"
  >
    <div
      className="ML-cat"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="closed-ML--category"
      >
        <span
          className="icon-database ml-icon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ModuleCategory  Library closed, category open matches snapshot 1`] = `
<div
  className="card closed-ML--category"
>
  <div
    className="module-category--wrapper"
  >
    <div
      className="ML-cat"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="closed-ML--category"
      >
        <span
          className="icon-undefined ml-icon"
        />
      </div>
    </div>
    <div
      className="ml-list-mini"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <Connect(DragSource(Module))
        addModule={[MockFunction]}
        dropModule={[MockFunction]}
        icon="add"
        id={88}
        isReadOnly={false}
        key="88"
        libraryOpen={false}
        name="First Module"
        setOpenCategory={[MockFunction]}
      />
      <Connect(DragSource(Module))
        addModule={[MockFunction]}
        dropModule={[MockFunction]}
        icon="url"
        id={101}
        isReadOnly={false}
        key="101"
        libraryOpen={false}
        name="Second Module"
        setOpenCategory={[MockFunction]}
      />
    </div>
  </div>
</div>
`;

exports[`ModuleCategory  Library open matches snapshot 1`] = `
<div
  className="card closed-ML--category"
>
  <div
    className="module-category--wrapper"
  >
    <div
      className="ML-cat"
    >
      <div
        className="category-container"
      >
        <span
          className="category-name"
        >
          Add Data
        </span>
      </div>
    </div>
    <div
      className="ml-list"
    >
      <Connect(DragSource(Module))
        addModule={[MockFunction]}
        dropModule={[MockFunction]}
        icon="add"
        id={88}
        isReadOnly={false}
        key="88"
        libraryOpen={true}
        name="First Module"
        setOpenCategory={[MockFunction]}
      />
      <Connect(DragSource(Module))
        addModule={[MockFunction]}
        dropModule={[MockFunction]}
        icon="url"
        id={101}
        isReadOnly={false}
        key="101"
        libraryOpen={true}
        name="Second Module"
        setOpenCategory={[MockFunction]}
      />
    </div>
  </div>
</div>
`;
